File: Python/Generate Parentheses.py
#!/usr/bin/env python


class Solution:
    # @param an integer
    # @return a list of string
    def generate(self, n, seq, st, left, right):
        if left == n and right == n:
            seq.append(st)
            return seq
        elif left == n:
            st1 = st + ')'
            seq = self.generate(n, seq, st1, left, right+1)
        elif left < n:
            if left > right:
                st1 = st + ')'
                seq = self.generate(n, seq, st1, left, right + 1)
            st1 = st + '('
            seq = self.generate(n, seq, st1, left + 1, right)
        return seq
    def generateParenthesis(self, n):
        seq = []
        st = ''
        return self.generate(n, seq, st, 0, 0)

File: C++/Generate Parentheses.cpp
class Solution {
public:
    void helper(vector<string> &res, string str, int x, int y, int n) {
        if (y == n) {
            res.push_back(str);
            return;
        }
        if (x == n) {
            string str1 = str + ")";
            helper(res, str1, x, y + 1, n);
        } else {
            string str1 = str + "(";
            helper(res, str1, x + 1, y, n);
            if (x > y) {
                str1 = str + ")";
                helper(res, str1, x, y + 1, n);
            }
        }
    }
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        helper(res, "", 0, 0, n);
        return res;
    }
};